mmdet/models/necks/bfp.py

```diff
@@ -0,0 +1,163 @@
+            assert num_outs >= self.num_ins - start_level
+        else:
+            # if end_level < inputs, no extra level is allowed
+            self.backbone_end_level = end_level
+            assert end_level <= len(in_channels)
+            assert num_outs == end_level - start_level
+        self.start_level = start_level
+        self.end_level = end_level
+        self.add_extra_convs = add_extra_convs
+        self.extra_convs_on_inputs = extra_convs_on_inputs
+
+        self.refine_level = refine_level
+        self.refine_type = refine_type
+
+        self.ops = []
+        self.rops = []
+        for i in range(self.start_level, self.backbone_end_level):
+            if i < self.refine_level:
+                self.ops.append(F.adaptive_max_pool2d)
+                self.rops.append(F.interpolate)
+            else:
+                self.ops.append(F.interpolate)
+                self.rops.append(F.adaptive_max_pool2d)
+
+        if self.refine_type == 'conv':
+            self.refine = ConvModule(
+                out_channels,
+                out_channels,
+                3,
+                padding=1,
+                conv_cfg=self.conv_cfg,
+                norm_cfg=self.norm_cfg,
+                activation=activation)
+        elif self.refine_type == 'non_local':
+            self.refine = NonLocalBlock2D(
+                out_channels,
+                reduction=1,
+                conv_cfg=self.conv_cfg,
+                norm_cfg=self.norm_cfg,
+                activation=activation)
+
+        # add extra conv layers (e.g., RetinaNet)
+        extra_levels = num_outs - self.backbone_end_level + self.start_level
+        if add_extra_convs and extra_levels >= 1:
+            self.extra_convs = nn.ModuleList()
+            for i in range(extra_levels):
+                if i == 0 and self.extra_convs_on_inputs:
+                    in_channels = self.in_channels[self.backbone_end_level - 1]
+                else:
+                    in_channels = out_channels
+                extra_conv = ConvModule(
+                    in_channels,
+                    out_channels,
+                    3,
+                    stride=2,
+                    padding=1,
+                    conv_cfg=self.conv_cfg,
+                    norm_cfg=self.norm_cfg,
+                    activation=self.activation,
+                    inplace=False)
+                self.extra_convs.append(extra_conv)
+
+    def init_weights(self):
+        for m in self.modules():
+            if isinstance(m, nn.Conv2d):
+                xavier_init(m, distribution='uniform')
+
+    def forward(self, inputs):
+        assert len(inputs) == len(self.in_channels)
+
+        ops_params = [
+            dict(output_size=inputs[self.refine_level].size()[2:])
+            if i < self.refine_level else dict(
+                size=inputs[self.refine_level].size()[2:], mode='nearest')
+            for i in range(self.start_level, self.backbone_end_level)
+        ]
```

Comment:
This list comprehension is too complicated for reading. I prefer to just use for loops to get integrated features.